### PR TITLE
support ceramic safe wallet

### LINKED ISSUE #21: Support any EVM chain multi-signature wallet

### PR DESCRIPTION
**Pull request by YoshiyukiSakura**
close #21

### CODE REVIEW

File: src/base/ceramicSafeWallet.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      return null;
+    }
+    // must sort by address
+    tx.confirmations.sort(
+      (
+        left: SafeMultisigConfirmationResponse,
+        right: SafeMultisigConfirmationResponse
+      ) => {
+        const leftAddress = left.owner.toUpperCase();
+        const rightAddress = right.owner.toUpperCase();
```
Comment (hujw77) [inline src/base/ceramicSafeWallet.ts]: `toLowerCase` is better.

Comment (YoshiyukiSakura) [reply src/base/ceramicSafeWallet.ts]: done

File: src/base/ceramicApiKit.ts [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                      dataDecoded: " "
+                      ethGasPrice: " "
+                      isSuccessful: false
+                      executionDate: "0000-00-00T00:00:0.000Z"
+                      refundReceiver: " "
+                      submissionDate: "${(new Date()).toISOString()}"
+                      transactionHash: "${safeTxHash}"
+                      confirmationsRequired: ${threshold}
+                  }
+                  clientMutationId: "${(new Date()).toISOString()}"
```
Comment (YoshiyukiSakura) [inline src/base/ceramicApiKit.ts]: check does it duplicate

File: src/base/ceramicApiKit.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                            signatureType
+                            submissionDate
+                            transactionHash
+                            confirmationType
+                        }
+                    }
+                }
+            }
+        `);
+      // @ts-ignore
```
Comment (YoshiyukiSakura) [inline src/base/ceramicApiKit.ts]: remove this ts-ignore

File: src/configure/configure.service.ts [hunk 4]
```diff
@@ -87,6 +87,8 @@ export interface BridgeInfo {
   safeWalletAddress: string | undefined;
   safeWalletUrl: string | undefined;
   safeWalletRole: string | undefined;
+  safeWalletType: "Gnosis" | "Ceramic" | undefined;
+  safeWalletCeramicUrl: string | undefined;
```
Comment (YoshiyukiSakura) [inline src/configure/configure.service.ts]: just use `safeWalletUrl`

File: src/base/ceramicApiKit.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
+import { ComposeClient } from "@composedb/client";
+import { SafeMultisigTransactionResponse } from "@safe-global/safe-core-sdk-types";
+import { ProposeTransactionProps } from "@safe-global/api-kit/dist/src/types/safeTransactionServiceTypes";
+import { definition } from "./ceramicModels";
+import { RuntimeCompositeDefinition } from "@composedb/types";
+
+export async function getComposeClient() {
```
Comment (YoshiyukiSakura) [inline src/base/ceramicApiKit.ts]: remove this wrap

File: src/base/ceramicApiKit.ts (hunk 2 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: yes it accepts any value

File: src/base/ceramicApiKit.ts (hunk 3 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: done

File: src/base/ceramicApiKit.ts (hunk 5 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: changed to a different style

File: src/base/ceramicSafeWallet.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  async proposeTransaction(
+    transactions: MetaTransactionData[],
+    isProposor: boolean,
+    chainId: bigint
+  ): Promise<TransactionPropose | null> {
+    this.safeSdk ?? (await this.connect(chainId));
+    const nonce = await this.safeSdk.getNonce();
+    const tx = await this.safeSdk.createTransaction({ transactions, options: { nonce } });
```
Comment (xiaoch05) [inline src/base/ceramicSafeWallet.ts]: Writing a nonce at this step does not work for transaction generation.
see https://github.com/safe-global/safe-core-sdk/blob/main/packages/protocol-kit/src/utils/transactions/utils.ts#L45

File: src/base/ceramicApiKit.ts [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+      const confirmations = confirmationsIndex.data.confirmationIndex.edges.map((edge) => edge.node).filter((confirmation) => {
+        const { signature } = confirmation;
+        const r = signature.slice(0, 66);
+        const s = `0x${signature.slice(66, 130)}`;
+        let v = parseInt(signature.slice(130, 132), 16);
+        if (v !== 27 && v !== 28) {
+          v = 27;
+        }
+        // normalize signature
+        const normalizedSignature = r + s.slice(2) + (v).toString(16).padStart(2, '0');
```
Comment (xiaoch05) [inline src/base/ceramicApiKit.ts]: Why is it necessary to separate and then merge this signature?

File: src/base/ceramicApiKit.ts [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        return ethers.verifyMessage(ethers.getBytes(safeTxHash), normalizedSignature).toLowerCase() === confirmation.owner.toLowerCase();
+      })
+      return {
+        ...transactionIndex.data.transactionIndex.edges[0].node,
+        confirmations
+      } as SafeMultisigTransactionResponse;
```
Comment (xiaoch05) [inline src/base/ceramicApiKit.ts]: What to do if there are duplicates?

File: src/base/ceramicSafeWallet.ts (hunk 6 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicSafeWallet.ts]: yes it does, 
so here the `options` data would be injected  https://github.com/safe-global/safe-core-sdk/blob/main/packages/protocol-kit/src/Safe.ts#L482, which specifies our passed in `nonce`

File: src/base/ceramicApiKit.ts (hunk 8 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: do you mean `confirmations` duplicates?

File: src/base/ceramicApiKit.ts [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    try {
+      const { ComposeClient } = await import('@composedb/client');
+      const composeClient = new ComposeClient({
+        ceramic: this.ceramicUrl,
+        //cannot import type from ESM only module in nest.js
+        // @ts-ignore
+        definition: definition,
+      });
+      const { fromString } = await import('uint8arrays/from-string');
+      const seedArray = fromString(this.privateKey, "base16");
```
Comment (xiaoch05) [inline src/base/ceramicApiKit.ts]: Is it safe to use the same private key as the safewallet account?

File: src/base/ceramicApiKit.ts [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        let v = parseInt(signature.slice(130, 132), 16);
+        if (v !== 27 && v !== 28) {
+          v = 27;
+        }
+        // normalize signature
+        const normalizedSignature = r + s.slice(2) + (v).toString(16).padStart(2, '0');
+        return ethers.verifyMessage(ethers.getBytes(safeTxHash), normalizedSignature).toLowerCase() === confirmation.owner.toLowerCase();
+      })
+      return {
+        ...transactionIndex.data.transactionIndex.edges[0].node,
```
Comment (xiaoch05) [inline src/base/ceramicApiKit.ts]: How to make sure the txdata is correct?

File: src/base/ceramicApiKit.ts (hunk 7 above)
Comment (xiaoch05) [reply src/base/ceramicApiKit.ts]: If you're dealing with v fields, ethers.js handles them accordingly.
https://github.com/ethers-io/ethers.js/blob/main/lib.esm/crypto/signature.js#L204

Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: in our case, `v` is `31`, the built-in normalize doesn't really work for us 
btw, ethersv5 does while ethersv6 not

File: src/base/ceramicApiKit.ts (hunk 8 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: and resolved, filtered here
https://github.com/helix-bridge/relayer/pull/30/files#diff-afa05a4785f248ff532b3a9c70ad3e9bbefcdd15491c95565b8e59e8546b187cR58

File: src/base/ceramicSafeWallet.ts (hunk 6 above)
Comment (xiaoch05) [reply src/base/ceramicSafeWallet.ts]: What I mean is that the following two implementations seem to work the same way.
1. const nonce = await this.safeSdk.getNonce();
    const tx = await this.safeSdk.createTransaction({ transactions, options: { nonce } });
2. const tx = await this.safeSdk.createTransaction({ transactions });

File: src/base/ceramicApiKit.ts (hunk 7 above)
Comment (xiaoch05) [reply src/base/ceramicApiKit.ts]: Our test case is 31, is it possible that for other chains the value is 30 or some other number?

File: src/base/ceramicApiKit.ts (hunk 9 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: very difficult to answer, 
on the surface, these packages doesn't send private to anywhere, just sign signatures,
but I can get your concern，let's me do research and see if threre's  more secured solution

File: src/base/ceramicApiKit.ts (hunk 7 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: I guess only way to find out is test， would let you know once I done

File: src/base/ceramicSafeWallet.ts (hunk 6 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicSafeWallet.ts]: correct，safeSDK already manages nonce for us
we can just read `nonce` from 
```
const tx = await this.safeSdk.createTransaction({ transactions, options: { nonce } });
tx.data.nonce
```

to me, I feel we should just use safeSdk internal nonce logic

File: src/base/ceramicApiKit.ts (hunk 10 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: validation would be the only answer, but this just reminds me of attack proof
what we can do if there are massive false data with the same `safeTxHash `, how to query the right one record 🤔

Comment (xiaoch05) [reply src/base/ceramicApiKit.ts]: I don't think we need to store the data in the remote storage, we just need to generate it locally(it's trustworthy) and compare it with safeTxHash, the current way of doing this is to get the first data in the remote storage, but there is no trustworthy source for this data.

Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: nice comment, 🤔 very interesting idea，we may just drop the data model `Transaction` and only use `Confirmation`, let me double confirm does the idea work

Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: I have done refactor, now it is reading TxData from local generation
@xiaoch05 and would like to know your suggestion, what do you think, let's remove Ceramic Transaction Model for good, or keep it, for potential later use?

File: src/base/ceramicApiKit.ts (hunk 9 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: let's use other key

File: src/base/ceramicSafeWallet.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      readyExecute: false,
+      safeTxHash: safeTxHash,
+      txData: "",
+      to: "",
+      value: BigInt(0),
+      operation: 0,
+      signatures: "",
+    };
+  }
+}
```
Comment (xiaoch05) [inline src/base/ceramicSafeWallet.ts]: The problem of duplicate codes is not completely solved.

File: src/base/ceramicApiKit.ts [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                        }
+                        clientMutationId: null
+                    }
+                ) {
+                    clientMutationId
+                }
+            }
+        `);
+
+    const transaction = await this.composeClient.executeQuery(`mutation CreateTransaction {
```
Comment (xiaoch05) [inline src/base/ceramicApiKit.ts]: Please delete this storage if it's not used.

Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: thank you! the comment answers this question
https://github.com/helix-bridge/relayer/pull/30#discussion_r1682020468

Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: done @xiaoch05

File: src/base/ceramicSafeWallet.ts (hunk 11 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicSafeWallet.ts]: refactored this `return` and also the `connect()` function
new they have very very few in common, which doesn't seem worthy further refactor

File: src/base/ceramicApiKit.ts [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+    try {
+      if (!this.composeClient) {
+        await this.connect();
+      }
+
+      const confirmationsIndex: ConfirmationIndexData = await this.composeClient.executeQuery(`
+            query ConfirmationIndex {
+                confirmationIndex(
+                    first: 99
```
Comment (xiaoch05) [inline src/base/ceramicApiKit.ts]: why this is 99, and does this number ensure that nothing goes wrong.

File: src/base/ceramicApiKit.ts (hunk 7 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: and it does change, 
31 on `base-sepolia`
32 on `sepolia`

File: src/base/ceramicApiKit.ts (hunk 13 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: just to higher the capacity, at the end of a day there's no absolute guarantee this query always works,
image someone is try to attack our service, and he sent 100 of fake signatures with the `safeTxHash`
which could not steal our money, but make the propose failed

personally, this hypothesis is low possibility to me, but set query limit to a reasonable large number makes me more confident about it running

File: src/base/ceramicApiKit.ts (hunk 7 above)
Comment (YoshiyukiSakura) [reply src/base/ceramicApiKit.ts]: working on find out more
these are only testnets supported by Safe Genosis

**Comment by xiaoch05** [comment]
leave one issue: https://github.com/helix-bridge/relayer/issues/35

File: src/base/ceramicApiKit.ts (hunk 13 above)
Comment (xiaoch05) [reply src/base/ceramicApiKit.ts]: https://github.com/helix-bridge/relayer/issues/35

File: src/relayer/relayer.service.ts [hunk 14]
```diff
@@ -261,8 +262,10 @@ export class RelayerService implements OnModuleInit {
           };
           if (config.safeWalletRole !== undefined) {
             let safeService: SafeService =
-              config.safeWalletType === "Ceramic"
+              config.safeWalletType === "ceramic"
                 ? new CeramicService(ceramicKey, config.safeWalletUrl)
+                : config.safeWalletType === "single"
```
Comment (YoshiyukiSakura) [inline src/relayer/relayer.service.ts:267]: maybe use `safeSdk.getThreshold() === 1`

**Pull request merged**